clang/include/clang/CIR/Dialect/IR/CIRAttrs.td

```diff
@@ -393,12 +396,88 @@ class ASTDecl<string name, string prefix, list<Trait> traits = []>
 
   // Enable verifier.
   let genVerifyDecl = 1;
-}
 
-def ASTFunctionDeclAttr : ASTDecl<"FunctionDecl", "fndecl">;
-def ASTVarDeclAttr : ASTDecl<"VarDecl", "vardecl">;
-def ASTRecordDeclAttr : ASTDecl<"RecordDecl", "recdecl">;
+  let extraClassDefinition = [{
+    ::mlir::Attribute $cppClass::parse(::mlir::AsmParser &parser,
+                                        ::mlir::Type type) {
+      // We cannot really parse anything AST related at this point
+      // since we have no serialization/JSON story.
+      return $cppClass::get(parser.getContext(), nullptr);
+    }
+
+    void $cppClass::print(::mlir::AsmPrinter &printer) const {
+      // Nothing to print besides the mnemonics.
+    }
+
+    LogicalResult $cppClass::verify(
+        ::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError,
+        }] # clang_name # [{ decl) {
+      return success();
+    }
+  }];
+}
 
+def ASTDeclAttr : ASTDecl<"Decl", "decl", [ASTDeclInterface]>;
+
+def ASTFunctionDeclAttr : ASTDecl<"FunctionDecl", "function.decl",
+  [ASTFunctionDeclInterface]
+>;
```

Comment:
Thanks for fixing most of the indentation stuff here.

Remaining one: we usually place the `>` right after the thing it closes, can you make this change for this and all the other ones?

Reply:
fixed in 74c93530a829f58866c5b0914f361cb3e0a53dc6